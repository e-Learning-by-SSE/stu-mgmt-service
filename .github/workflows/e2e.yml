name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
