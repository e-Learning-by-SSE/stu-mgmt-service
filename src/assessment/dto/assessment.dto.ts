import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
