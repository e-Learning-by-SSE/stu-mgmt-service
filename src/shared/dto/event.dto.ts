export class EventDto {
	event: string;
	timestamp: Date;
	payload?: Record<string, unknown>;
}
