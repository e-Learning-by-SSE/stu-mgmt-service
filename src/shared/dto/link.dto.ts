export class LinkDto {
	url: string;
	name: string;
}
