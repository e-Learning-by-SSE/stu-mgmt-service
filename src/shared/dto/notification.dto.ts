import { AssignmentId } from "../../course/entities/assignment.entity";
import { CourseId } from "../../course/entities/course.entity";
import { GroupId } from "../../course/entities/group.entity";
import { Event } from "../../course/events/events";
import { UserId } from "../entities/user.entity";

export class NotificationDto {
	event: Event;
	courseId: CourseId;
	assignmentId?: AssignmentId;
	groupId?: GroupId;
	userId?: UserId;
	payload?: Record<string, unknown>;
}
