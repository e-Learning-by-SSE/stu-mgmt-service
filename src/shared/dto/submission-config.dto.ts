import { IsNotEmpty } from "class-validator";

export class SubmissionConfigDto {
	@IsNotEmpty()
    tool: string;
    @IsNotEmpty()
	config: string;
}
