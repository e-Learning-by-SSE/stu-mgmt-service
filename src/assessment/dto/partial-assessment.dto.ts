import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
