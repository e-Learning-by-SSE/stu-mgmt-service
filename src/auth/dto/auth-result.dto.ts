import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
