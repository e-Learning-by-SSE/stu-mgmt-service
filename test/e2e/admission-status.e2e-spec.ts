import { AdmissionStatusDto } from "../../src/admission-status/dto/admission-status.dto";
import { PointsOverviewDto } from "../../src/admission-status/dto/points-overview.dto";
import { USER_STUDENT_JAVA } from "../mocks/users.mock";
import { TestSetup } from "../utils/e2e";

describe("Admission Status", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
		await setup.dbMockService.createAll();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	const courseId = "java-wise1920";

	it("[GET] /courses/{courseId}/admission-status", () => {
		return setup
			.requestWithAuth("get", `/courses/${courseId}/admission-status`)
			.expect(200)
			.expect(({ body }) => {
				const result = body as AdmissionStatusDto[];
				expect(result).toHaveLength(4);
				expect(result).toMatchSnapshot();
			});
	});

	it("[GET] /courses/{courseId}/admission-status/{userId}", () => {
		return setup
			.requestWithAuth("get", `/courses/${courseId}/admission-status/${USER_STUDENT_JAVA.id}`)
			.expect(200)
			.expect(({ body }) => {
				const result = body as AdmissionStatusDto;
				expect(result).toMatchSnapshot();
			});
	});

	it("[GET] /courses/{courseId}/admission-status/overview", () => {
		return setup
			.requestWithAuth("get", `/courses/${courseId}/admission-status/overview`)
			.expect(200)
			.expect(({ body }) => {
				const result = body as PointsOverviewDto;
				expect(result.assignments).toHaveLength(2); // 2 evaluated assignments
				expect(result.results).toHaveLength(4); // 4 students
				expect(result).toMatchSnapshot();
			});
	});

	it("[GET] /courses/{courseId}/admission-status/overview/{userId}", () => {
		return setup
			.requestWithAuth(
				"get",
				`/courses/${courseId}/admission-status/overview/${USER_STUDENT_JAVA.id}`
			)
			.expect(200)
			.expect(({ body }) => {
				const result = body as PointsOverviewDto;
				expect(result.assignments).toHaveLength(2); // 2 evaluated assignments
				expect(result.results).toHaveLength(1);
				expect(result).toMatchSnapshot();
			});
	});
});
