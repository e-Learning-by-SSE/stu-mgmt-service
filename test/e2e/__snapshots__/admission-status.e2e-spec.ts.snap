// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admission Status [GET] /courses/{courseId}/admission-status 1`] = `
Array [
  Object {
    "fulfillsAdmissionCriteria": false,
    "hasAdmission": false,
    "hasAdmissionFromPreviousSemester": false,
    "participant": Object {
      "displayName": "Max Mustermann",
      "email": "max.mustermann@test.com",
      "matrNr": 123456,
      "role": "STUDENT",
      "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
      "username": "mmustermann",
    },
    "results": Array [
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 75,
        "achievedPoints": 75,
        "passed": true,
      },
      Object {
        "_assignmentType": "TESTAT",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 25,
        "achievedPoints": 25,
        "passed": false,
      },
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": true,
      },
    ],
  },
  Object {
    "fulfillsAdmissionCriteria": false,
    "hasAdmission": false,
    "hasAdmissionFromPreviousSemester": false,
    "participant": Object {
      "displayName": "Hans Peter",
      "email": "hans.peter@test.com",
      "matrNr": 111111,
      "role": "STUDENT",
      "userId": "40f59aad-7473-4455-a3ea-1214f19b2565",
      "username": "hpeter",
    },
    "results": Array [
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 75,
        "achievedPoints": 75,
        "passed": true,
      },
      Object {
        "_assignmentType": "TESTAT",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": false,
      },
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": true,
      },
    ],
  },
  Object {
    "fulfillsAdmissionCriteria": false,
    "hasAdmission": false,
    "hasAdmissionFromPreviousSemester": false,
    "participant": Object {
      "displayName": "elshar",
      "email": "elshar@test.com",
      "matrNr": 777777,
      "role": "STUDENT",
      "userId": "3e52e822-4ebc-49c3-94e2-06ba447b5d1f",
      "username": "elshar",
    },
    "results": Array [
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": false,
      },
      Object {
        "_assignmentType": "TESTAT",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": false,
      },
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
        "achievedPercent": 50,
        "achievedPoints": 1,
        "passed": true,
      },
    ],
  },
  Object {
    "fulfillsAdmissionCriteria": false,
    "hasAdmission": false,
    "hasAdmissionFromPreviousSemester": false,
    "participant": Object {
      "displayName": "kunold",
      "email": "kunold@test.com",
      "matrNr": 888888,
      "role": "STUDENT",
      "userId": "24b4ec8c-dd8a-4bd6-843c-2c412fb1d5b7",
      "username": "kunold",
    },
    "results": Array [
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": false,
      },
      Object {
        "_assignmentType": "TESTAT",
        "_rule": "REQUIRED_PERCENT_OVERALL",
        "achievedPercent": 0,
        "achievedPoints": 0,
        "passed": false,
      },
      Object {
        "_assignmentType": "HOMEWORK",
        "_rule": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
        "achievedPercent": 50,
        "achievedPoints": 1,
        "passed": true,
      },
    ],
  },
]
`;

exports[`Admission Status [GET] /courses/{courseId}/admission-status/{userId} 1`] = `
Object {
  "fulfillsAdmissionCriteria": false,
  "hasAdmission": false,
  "hasAdmissionFromPreviousSemester": false,
  "participant": Object {
    "displayName": "Max Mustermann",
    "email": "max.mustermann@test.com",
    "matrNr": 123456,
    "role": "STUDENT",
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
    "username": "mmustermann",
  },
  "results": Array [
    Object {
      "_assignmentType": "HOMEWORK",
      "_rule": "REQUIRED_PERCENT_OVERALL",
      "achievedPercent": 75,
      "achievedPoints": 75,
      "passed": true,
    },
    Object {
      "_assignmentType": "TESTAT",
      "_rule": "REQUIRED_PERCENT_OVERALL",
      "achievedPercent": 25,
      "achievedPoints": 25,
      "passed": false,
    },
    Object {
      "_assignmentType": "HOMEWORK",
      "_rule": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
      "achievedPercent": 0,
      "achievedPoints": 0,
      "passed": true,
    },
  ],
}
`;

exports[`Admission Status [GET] /courses/{courseId}/admission-status/overview 1`] = `
Object {
  "assignments": Array [
    Object {
      "collaboration": "GROUP_OR_SINGLE",
      "endDate": "2020-05-10T22:00:00.000Z",
      "id": "c2bc4aa4-6207-11ea-bc55-0242ac130003",
      "name": "Test_Assignment 04 (Java)",
      "points": 100,
      "startDate": "2020-05-03T22:00:00.000Z",
      "state": "EVALUATED",
      "type": "HOMEWORK",
    },
    Object {
      "collaboration": "SINGLE",
      "endDate": "2020-04-26T22:00:00.000Z",
      "id": "75b799a1-a406-419b-a448-909aa3d34afa",
      "name": "Test_Assignment 07 (Java) Testat Evaluated",
      "points": 100,
      "startDate": "2020-04-19T22:00:00.000Z",
      "state": "EVALUATED",
      "type": "TESTAT",
    },
  ],
  "results": Array [
    Object {
      "achievedPoints": Array [
        75,
        25,
      ],
      "assessmentIds": Array [
        "8f60f844-4129-48a4-a625-7a74c7defd0d",
        "932c7bd8-2338-4e60-955a-39da5f858212",
      ],
      "student": Object {
        "displayName": "Max Mustermann",
        "email": "max.mustermann@test.com",
        "matrNr": 123456,
        "role": "STUDENT",
        "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
        "username": "mmustermann",
      },
    },
    Object {
      "achievedPoints": Array [
        75,
        null,
      ],
      "assessmentIds": Array [
        "8f60f844-4129-48a4-a625-7a74c7defd0d",
        null,
      ],
      "student": Object {
        "displayName": "Hans Peter",
        "email": "hans.peter@test.com",
        "matrNr": 111111,
        "role": "STUDENT",
        "userId": "40f59aad-7473-4455-a3ea-1214f19b2565",
        "username": "hpeter",
      },
    },
    Object {
      "achievedPoints": Array [
        null,
        null,
      ],
      "assessmentIds": Array [
        null,
        null,
      ],
      "student": Object {
        "displayName": "elshar",
        "email": "elshar@test.com",
        "matrNr": 777777,
        "role": "STUDENT",
        "userId": "3e52e822-4ebc-49c3-94e2-06ba447b5d1f",
        "username": "elshar",
      },
    },
    Object {
      "achievedPoints": Array [
        null,
        null,
      ],
      "assessmentIds": Array [
        null,
        null,
      ],
      "student": Object {
        "displayName": "kunold",
        "email": "kunold@test.com",
        "matrNr": 888888,
        "role": "STUDENT",
        "userId": "24b4ec8c-dd8a-4bd6-843c-2c412fb1d5b7",
        "username": "kunold",
      },
    },
  ],
}
`;

exports[`Admission Status [GET] /courses/{courseId}/admission-status/overview/{userId} 1`] = `
Object {
  "assignments": Array [
    Object {
      "collaboration": "GROUP_OR_SINGLE",
      "endDate": "2020-05-10T22:00:00.000Z",
      "id": "c2bc4aa4-6207-11ea-bc55-0242ac130003",
      "name": "Test_Assignment 04 (Java)",
      "points": 100,
      "startDate": "2020-05-03T22:00:00.000Z",
      "state": "EVALUATED",
      "type": "HOMEWORK",
    },
    Object {
      "collaboration": "SINGLE",
      "endDate": "2020-04-26T22:00:00.000Z",
      "id": "75b799a1-a406-419b-a448-909aa3d34afa",
      "name": "Test_Assignment 07 (Java) Testat Evaluated",
      "points": 100,
      "startDate": "2020-04-19T22:00:00.000Z",
      "state": "EVALUATED",
      "type": "TESTAT",
    },
  ],
  "results": Array [
    Object {
      "achievedPoints": Array [
        75,
        25,
      ],
      "assessmentIds": Array [
        "8f60f844-4129-48a4-a625-7a74c7defd0d",
        "932c7bd8-2338-4e60-955a-39da5f858212",
      ],
      "student": Object {
        "displayName": "Max Mustermann",
        "email": "max.mustermann@test.com",
        "matrNr": 123456,
        "role": "STUDENT",
        "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
        "username": "mmustermann",
      },
    },
  ],
}
`;
